server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java

```diff
@@ -121,38 +136,94 @@ private CurrentWsResponse toWsResponse(DbSession dbSession, String userLogin) {
+      return organizationHomepage(dbSession, user);
+    }
+
+    return of(CurrentWsResponse.Homepage.newBuilder()
+      .setType(CurrentWsResponse.HomepageType.valueOf(user.getHomepageType()))
+      .build());
+  }
+
+  private Optional<CurrentWsResponse.Homepage> projectHomepage(DbSession dbSession, UserDto user) {
+    Optional<ComponentDto> projectOptional = ofNullable(dbClient.componentDao().selectByUuid(dbSession, of(user.getHomepageParameter()).orElse(EMPTY)).orNull());
+    if (shouldCleanProjectHomepage(projectOptional)) {
+      cleanUserHomepageInDb(dbSession, user);
+      return empty();
     }
-    String homepageParameter = getHomepageParameter(dbSession, user.getHomepageType(), user.getHomepageParameter());
+
     CurrentWsResponse.Homepage.Builder homepage = CurrentWsResponse.Homepage.newBuilder()
-      .setType(CurrentWsResponse.HomepageType.valueOf(user.getHomepageType()));
-    setNullable(homepageParameter, homepage::setParameter);
-    return homepage.build();
+      .setType(CurrentWsResponse.HomepageType.valueOf(user.getHomepageType()))
+      .setComponent(projectOptional.get().getKey());
+    setNullable(projectOptional.get().getBranch(), homepage::setBranch);
+    return of(homepage.build());
   }
 
-  @CheckForNull
-  private String getHomepageParameter(DbSession dbSession, String homepageType, String homepageParameter) {
-    if (PROJECT.toString().equals(homepageType)) {
-      return dbClient.componentDao().selectByUuid(dbSession, homepageParameter)
-        .transform(ComponentDto::getKey)
-        .or(() -> {
-          throw new IllegalStateException(format("Unknown component '%s' for homepageParameter", homepageParameter));
-        });
+  private boolean shouldCleanProjectHomepage(Optional<ComponentDto> projectOptional) {
+    return !projectOptional.isPresent() || !userSession.hasComponentPermission(USER, projectOptional.get());
+  }
+
+  private Optional<CurrentWsResponse.Homepage> applicationAndPortfolioHomepage(DbSession dbSession, UserDto user) {
+    Optional<ComponentDto> componentOptional = ofNullable(dbClient.componentDao().selectByUuid(dbSession, of(user.getHomepageParameter()).orElse(EMPTY)).orNull());
+    if (shouldCleanApplicationOrPortfolioHomepage(componentOptional)) {
+      cleanUserHomepageInDb(dbSession, user);
+      return empty();
     }
-    if (ORGANIZATION.toString().equals(homepageType)) {
-      return dbClient.organizationDao().selectByUuid(dbSession, homepageParameter)
-        .map(OrganizationDto::getKey)
-        .orElseThrow(() -> new IllegalStateException(format("Unknown organization '%s' for homepageParameter", homepageParameter)));
+
+    return of(CurrentWsResponse.Homepage.newBuilder()
+      .setType(CurrentWsResponse.HomepageType.valueOf(user.getHomepageType()))
+      .setComponent(componentOptional.get().getKey())
+      .build());
+  }
+
+  private boolean shouldCleanApplicationOrPortfolioHomepage(Optional<ComponentDto> componentOptional) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Specify a "ComponentDto" parameter instead. ([squid:S3553](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1185%2Fhomepage_on_sonarqube&issues=AWHdP9XFw6XGHKvPuY58&open=AWHdP9XFw6XGHKvPuY58)

[ISSUE_KEY]: AWHdP9XFw6XGHKvPuY58